### ISSUE TITLE
cutorch install failed

### ISSUE THREAD

**Issue by kzachery** [torch/cutorch#175]
I tried to install cutorch, but it failed to build. Can you tell me what dependences I am missing? The following is the command I used for the install and the output listing the failure.

$ luarocks install cutorch

Scanning dependencies of target THC
[ 67%] [ 67%] [ 70%] [ 73%] Building C object lib/THC/CMakeFiles/THC.dir/THCAllocator.c.o
Building C object lib/THC/CMakeFiles/THC.dir/THCStorage.c.o
Building C object lib/THC/CMakeFiles/THC.dir/THCGeneral.c.o
Building C object lib/THC/CMakeFiles/THC.dir/THCStorageCopy.c.o
/tmp/luarocks_cutorch-scm-1-1359/cutorch/lib/THC/THCStorage.c:3:22: fatal error: THAtomic.h: No such file or directory
 #include "THAtomic.h"
                      ^
compilation terminated.
[ 76%] Building C object lib/THC/CMakeFiles/THC.dir/THCTensor.c.o
[ 79%] Building C object lib/THC/CMakeFiles/THC.dir/THCTensorCopy.c.o
/tmp/luarocks_cutorch-scm-1-1359/cutorch/lib/THC/THCTensor.c:4:22: fatal error: THAtomic.h: No such file or directory
 #include "THAtomic.h"
                      ^
compilation terminated.
make[2]: *** [lib/THC/CMakeFiles/THC.dir/THCStorage.c.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [lib/THC/CMakeFiles/THC.dir/THCTensor.c.o] Error 1
make[1]: *** [lib/THC/CMakeFiles/THC.dir/all] Error 2
make: *** [all] Error 2

Error: Build error: Failed building.

Many Thanks,
kzachery

**Comment by szagoruyko** [comment]
update torch first:
luarocks install torch
luarocks install cutorch

**Comment by kzachery** [comment]
Many Thanks!

**Comment by yossibiton** [comment]
I get exactly the same same error, although installing the most recent torch package before.
Any solution ?

**Comment by yossibiton** [comment]
Im using local install, using torch distrohttps://github.com/torch/distro

1. Running update.sh to make sure all submodules are nost updated
2. All installations went well, till i get to cutorch

**Comment by kzachery** [comment]
The following worked for me:

luarocks install torch
luarocks install cutorch

**Comment by yossibiton** [comment]
i see, i'm doing pretty same thing:
1. git clone https://github.com/torch/torch7
2. cd torch7
3. luarocks make rocks/torch-scm-1.rockspec
the same thing with cutorch.

i can't use luarocks install because i have no access to github on the target server.

**Comment by soumith** [comment]
@yossibiton it's likely that you have two versions of torch, a global but older version, a local but newer version. When you install cutorch, it is likely picking up the global-old version of headers from torch7

**Comment by yossibiton** [comment]
Absolutely right. I have an older global installation but now want to
switch to the local one (while keeping the global installation for other
users).

For the local installation im using torch distro, and thus the luarocks
built with this package (distro/install/bin/luarocks). This one should be
associated with the local install, isn't it ?
What is thr simplest way to force cutorch  refer the newer torch ?
thanks

**Comment by fmassa** [comment]
@yossibiton if you want to keep the global install, then instead of calling `luarocks make rocks/torch-scm-1.rockspec` try to do `~/torch/install/bin/luarocks make rocks/torch-scm-1.rockspec` (I supposed that you used the default path for the local installation of torch), and then the same thing for `cutorch`.

**Comment by yossibiton** [comment]
i did used '~/torch/install/bin/luarocks' and not 'luarocks' only.
surprisingly it did linked with the global torch however.

On Wed, Jun 17, 2015 at 4:38 PM, Francisco Massa <notifications@github.com>
wrote:

> @yossibiton <https://github.com/yossibiton> if you want to keep the
> global install, then instead of calling luarocks make
> rocks/torch-scm-1.rockspec try to do ~/torch/install/bin/luarocks make
> rocks/torch-scm-1.rockspec (I supposed that you used the default path for
> the local installation of torch), and then the same thing for cutorch.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/torch/cutorch/issues/175#issuecomment-112802402>.
>



-- 
Yossi Biton

**Comment by yossibiton** [comment]
anyone one how to force cutorch installation link with the right torch package ?

**Comment by yossibiton** [comment]
UPDATE :
I decided to remove any 
after @soumith assumption, i decided to remove the global torch installation i had (i hope i removed right all leftovers) and then install the updated of torch (running install.sh from https://github.com/torch/distro. 
Unfortunately, it still faild when reaching cutorch. 
@soumith, do you have a new idea to save my day ?

By the way, when i'm trying to install an older version of cutorch, which i cloned few months ago it succeed. (i'm using offline installation by first cloning the git repository, and then using luarocks make).

**Comment by soumith** [comment]
@yossibiton use the script clean-old.sh from the ezinstall repo to clean the old install. Afterwards give this a try.

**Comment by yossibiton** [comment]
Thank you @soumith .
That solved the problem (together with running clean.sh from the distro repo).
This issue can be closed.

**Comment by mendadala** [comment]
Can anyone help me? I am getting this error:(!

MARSs-MBP:must-cnn neptune$ th main.lua
warning: '?.split' will not override existing symbol
{
  gpu : 1
  datadir : "./data/4Protein/data"
  task : "absolute.lab"
  dataset : "./data/4Protein"
  kernelsize : 5
  hiddenunit : 1024
  nclass : 2
  hashdir : "./data/4Protein/hash"
  cuda : false
  nonlinearity : "relu"
  poolingsize : 2
  embedsize : 15
}
absolute.lab.tag.dat	
Start loading data
finish loading data	
nn.ClassNLLCriterion
nn.Sequential {
  [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> (7) -> (8) -> (9) -> (10) -> (11) -> (12) -> output]
  (1): nn.LookupTable
  (2): nn.Dropout(0.100000)
  (3): nn.UpDim
  (4): nn.ConcatTable {
    input
      |`-> (1): nn.SpatialZeroPadding(l=0, r=0, t=2, b=4)
       `-> (2): nn.SpatialZeroPadding(l=0, r=0, t=1, b=5)
       ... -> output
  }
  (5): nn.JoinTable
  (6): nn.TemporalConvolution
  (7): nn.ReLU
  (8): nn.TemporalMaxPooling
  (9): nn.Dropout(0.300000)
  (10): nn.TemporalZip
  (11): nn.TemporalConvolution
  (12): nn.LogSoftMax
}
nn.ClassNLLCriterion
Epoch 0: training...
/Users/neptune/torch/install/bin/luajit: /Users/neptune/torch/install/share/lua/5.1/nn/Container.lua:67: p: 0ms         
In 12 module of nn.Sequential:
/Users/neptune/torch/install/share/lua/5.1/nn/THNN.lua:110: input and gradOutput shapes do not match: input [292 x 2], gradOutput [291 x 2] at /tmp/luarocks_nn-scm-1-4851/nn/lib/THNN/generic/LogSoftMax.c:78
stack traceback:
	[C]: in function 'v'
	/Users/neptune/torch/install/share/lua/5.1/nn/THNN.lua:110: in function 'LogSoftMax_updateGradInput'
	...rs/neptune/torch/install/share/lua/5.1/nn/LogSoftMax.lua:12: in function 'updateGradInput'
	/Users/neptune/torch/install/share/lua/5.1/nn/Module.lua:31: in function </Users/neptune/torch/install/share/lua/5.1/nn/Module.lua:29>
	[C]: in function 'xpcall'
	/Users/neptune/torch/install/share/lua/5.1/nn/Container.lua:63: in function 'rethrowErrors'
	...rs/neptune/torch/install/share/lua/5.1/nn/Sequential.lua:84: in function 'backward'
	./Trainer.lua:93: in function 'opfunc'
	/Users/neptune/torch/install/share/lua/5.1/optim/sgd.lua:44: in function 'optimMethod'
	./Trainer.lua:108: in function 'train'
	main.lua:92: in function 'main'
	main.lua:125: in main chunk
	[C]: in function 'dofile'
	...tune/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk
	[C]: at 0x010ccf0bd0

WARNING: If you see a stack trace below, it doesn't point to the place where this error occurred. Please use only the one above.
stack traceback:
	[C]: in function 'error'
	/Users/neptune/torch/install/share/lua/5.1/nn/Container.lua:67: in function 'rethrowErrors'
	...rs/neptune/torch/install/share/lua/5.1/nn/Sequential.lua:84: in function 'backward'
	./Trainer.lua:93: in function 'opfunc'
	/Users/neptune/torch/install/share/lua/5.1/optim/sgd.lua:44: in function 'optimMethod'
	./Trainer.lua:108: in function 'train'
	main.lua:92: in function 'main'
	main.lua:125: in main chunk
	[C]: in function 'dofile'
	...tune/torch/install/lib/luarocks/rocks/trepl/scm-1/bin/th:150: in main chunk
	[C]: at 0x010ccf0bd0
